packages/grafana-ui/src/components/Table/TableCell.tsx

```diff
@@ -0,0 +1,60 @@
+import React, { FC } from 'react';
+import { Cell, CellProps, TableCellProps } from 'react-table';
+import { Field } from '@grafana/data';
+import { BackgroundColoredCell, DefaultCell } from './DefaultCell';
+import { BarGaugeCell } from './BarGaugeCell';
+import { getTextAlign } from './utils';
+import { TableCellDisplayMode, TableFieldOptions, TableFilterActionCallback } from './types';
+import { TableStyles } from './styles';
+
+interface Props {
+  cell: Cell;
+  field: Field;
+  tableStyles: TableStyles;
+  onCellClick?: TableFilterActionCallback;
+}
+
+export const TableCell: FC<Props> = ({ cell, field, tableStyles, onCellClick }) => {
+  const filterable = field.config.filterable;
+  const cellProps = cell.getCellProps();
+
+  let onClick: ((event: React.SyntheticEvent) => void) | undefined = undefined;
+  if (filterable && onCellClick) {
+    if (cellProps.style) {
+      cellProps.style.cursor = 'pointer';
+    }
+
+    onClick = () => onCellClick(cell.column.Header as string, cell.value);
+  }
+  const fieldTextAlign = getTextAlign(field);
+  if (fieldTextAlign && cellProps.style) {
+    cellProps.style.textAlign = fieldTextAlign;
+  }
+
+  cell.column.Cell = getCellDisplayType(cellProps, field, tableStyles);
+
+  return (
+    <div {...cellProps} onClick={onClick}>
+      {cell.render('Cell')}
+    </div>
+  );
+};
+
+function getCellDisplayType(cellProps: TableCellProps, field: Field, tableStyles: TableStyles) {
+  let Cell = (cellProps: CellProps<any>) => <DefaultCell field={field} {...cellProps} tableStyles={tableStyles} />;
+
+  const fieldTableOptions = (field.config.custom || {}) as TableFieldOptions;
+  switch (fieldTableOptions.displayMode) {
+    case TableCellDisplayMode.ColorBackground:
+      Cell = (cellProps: CellProps<any>) => (
```

Comment:
As @torkelo noticed this will create a new component for every cell rendered what can be a performance issue. What we can do instead is to pass the additional props (field, tableStyles) when the cell is rendered:

```
cell.render('Cell', {field, tableStyles})
```

Reply:
so basically move the switch statement back to getColumns